File: python/scripts/update_schema.py
"""
YOU DO NOT NEED TO RUN THIS SCRIPT UNLESS THE GRAPHQL API HAS CHANGED

Introspect the Reports GraphQL API and update the local schema file.

1. Under the launch menu select "⚠️🐍 Python: Update Schema"
2. Commit the changes to the schema file at `pyreports/graphql/rs-reports.schema.graphql`

"""

from pathlib import Path

import questionary
import requests
from graphql import build_client_schema, get_introspection_query, print_schema

API_URLS = {
    "staging": "https://api.reports.riverscapes.net/staging",
    "production": "https://api.reports.riverscapes.net",
}

SCHEMA_PATH = Path(__file__).parent.parent / "pyreports" / "graphql" / "rs-reports.schema.graphql"


def main():
    stage = questionary.select(
        "Which API stage to introspect?",
        choices=list(API_URLS.keys()),
        default="staging",
    ).ask()
    if not stage:
        return

    url = API_URLS[stage]
    print(f"Introspecting {stage} API at {url} ...")

    response = requests.post(
        url,
        json={"query": get_introspection_query()},
        headers={"Content-Type": "application/json"},
        timeout=30,
    )
    response.raise_for_status()
    result = response.json()

    if "errors" in result:
        raise SystemExit(f"Introspection failed: {result['errors']}")

    schema = build_client_schema(result["data"])
    sdl = print_schema(schema)

    SCHEMA_PATH.write_text(sdl + "\n", encoding="utf-8")
    print(f"Schema written to {SCHEMA_PATH}")


if __name__ == "__main__":
    main()

File: typescript/scripts/updateSchema.ts
/**
 * Introspect the Reports GraphQL API and update the local schema file.
 */

import * as fs from 'node:fs'
import * as path from 'node:path'
import { select } from '@inquirer/prompts'
import { type IntrospectionQuery, getIntrospectionQuery, buildClientSchema, printSchema } from 'graphql'

const API_URLS: Record<string, string> = {
  staging: 'https://api.reports.riverscapes.net/staging',
  production: 'https://api.reports.riverscapes.net',
}

const SCHEMA_PATH = path.resolve(import.meta.dirname, '..', 'src', 'graphql', 'rs-reports.schema.graphql')

async function main(): Promise<void> {
  const stage = await select({
    message: 'Which API stage to introspect?',
    choices: Object.keys(API_URLS).map((s) => ({ name: s, value: s })),
    default: 'staging',
  })

  const url = API_URLS[stage]
  console.log(`Introspecting ${stage} API at ${url} ...`)

  const response = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ query: getIntrospectionQuery() }),
  })

  if (!response.ok) {
    throw new Error(`HTTP ${response.status} from API`)
  }

  const result = (await response.json()) as { data?: Record<string, unknown>; errors?: unknown[] }

  if (result.errors) {
    throw new Error(`Introspection failed: ${JSON.stringify(result.errors)}`)
  }

  const schema = buildClientSchema(result.data as unknown as IntrospectionQuery)
  const sdl = printSchema(schema)

  fs.writeFileSync(SCHEMA_PATH, sdl + '\n', 'utf-8')
  console.log(`Schema written to ${SCHEMA_PATH}`)
}

main().catch(console.error)
